### PR TITLE
drivers: enhance check for mount fs type

### PR DESCRIPTION
**Pull request by giuseppe**
when checking that a mount has a specific file system type, also
validate that it is different than its parent directory.

This helps when using virtiofs as the underlying file system since the
check used for fuse-overlayfs would fail since they both use FUSE.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: drivers/driver_linux.go [hunk 1]
```diff
@@ -168,5 +169,16 @@ func Mounted(fsType FsMagic, mountPath string) (bool, error) {
...
+	}
+	// it is already the root
+	if mountPath == "/" {
+		return true, nil
+	}
+
+	if err := unix.Statfs(filepath.Dir(mountPath), &buf); err != nil {
+		return true, err
+	}
+	return FsMagic(buf.Type) != fsType, nil
```
[the remark this answers is not in the document]
Comment (giuseppe) [reply drivers/driver_linux.go:183]: `buf` is overwritten with the data from the parent directory, the check is on a different directory now

[the remark this answers is not in the document]
Comment (rhatdan) [reply drivers/driver_linux.go:183]: Yup, now I see it it. I did not read the statfs line it was modifying buf as a side effect.
LGTM

**Comment by nalind** [comment]
> when checking that a mount has a specific file system type, also validate that it is different than its parent directory.

If one FUSE filesystem is mounted somewhere below a different FUSE filesystem, why would we want this function to start to return `false` where it returned `true` before?

> This helps when using virtiofs as the underlying file system since the check used for fuse-overlayfs would fail since they both use FUSE.

Can you elaborate on what this helps with?

**Comment by rhvgoyal** [comment]
So we are trying to run "buildah bud" inside a kata container and that fails with error.
`
kernel does not support overlay fs: unable to create kernel-style whiteout: operation not permitted
WARN[0000] failed to shutdown storage: "kernel does not support overlay fs: unable to create kernel-style whiteout:operation not permitted
`

Kata container rootfs is virtiofs which is backed by overlayfs on host.

We were expecting that fuse-overlayfs will be used on top of virtiofs and that should work out of the box. But that's not happening for some reason.

I don't understand this code fully but giuseppe pointed me to this code.

So question is, what's wrong with running a fuse filesystem on top another and why filesystem type matters. And why we are not switching to using fuse-overlay.

Kernel overlayfs will not work as it is because whiteout creation fails on top of host overlayfs. fuse-overlayfs can work as it will switch to regular file starting with .wh as whiteout.

**Comment by nalind** [comment]
What's the error that this change helps us avoid?  Is the overlay driver getting confused and "detecting" that a fuse-overlayfs mount is already mounted when it isn't, or is it something else?

**Comment by rhvgoyal** [comment]
No idea. It will be good to have more explanation of current behavior and why that current behavior is not working on top of virtiofs.

**Comment by rhatdan** [comment]
We have a report where podman with user namespace is showing underlying files owned by nobody instead of root.

**Comment by giuseppe** [comment]
> What's the error that this change helps us avoid? Is the overlay driver getting confused and "detecting" that a fuse-overlayfs mount is already mounted when it isn't, or is it something else?

the error is that the entire `/` is mounted on top of a FUSE file system.  The current check will always return true, so we don't do any `fuse-overlayfs` mount and we don't do any container mount (thus the rootfs will be empty when the container starts).

The proposed fix instead also checks that the target directory is a new mount point, different than its parent.

**Comment by gkurz** [comment]
I've rebuilt `buildah` incorporating this change and tested it inside a kata container running on openshift. This fixes the issue we're currently hitting in https://issues.redhat.com/browse/KATA-1278 .

Thanks @giuseppe !

File: drivers/driver_linux.go [hunk 2]
```diff
@@ -165,8 +166,22 @@ func (c *defaultChecker) IsMounted(path string) bool {
...
+	// it is already the root
+	if mountPath == "/" {
+		return true, nil
+	}
+	if err := unix.Statfs(filepath.Dir(mountPath), &bufParent); err != nil {
+		return true, err
+	}
+	// check that the parent directory is on a different mount
+	// than mountPath
+	return bufParent.Fsid != buf.Fsid, nil
```
Comment (nalind) [inline drivers/driver_linux.go:186]: I like the idea of comparing information about the two directories, but the kernel currently always returns an fsid of 0 for fuse filesystems, so this comparison will always be `false`.  The `st_dev` value returned when we `syscall.Stat()` the directories, though, should be unique for different FUSE mounts.

Comment (giuseppe) [reply drivers/driver_linux.go]: good idea, I've amended the patch

**Comment by nalind** [comment]
/approve
LGTM

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**